### ISSUE TITLE
Add Tiberian Sun/RA2 light flashes from weapons fire and explosions

### ISSUE THREAD

**Issue by henskelion** [OpenRA/OpenRA#18658]
In RA2 and TS, some weapons fire and warhead impacts would create a small flash of light for a moment. 

There is some documentation on how these light flashes worked that can be found here: https://www.modenc.renegadeprojects.com/Bright

A tutorial also exists for recreating this in RA2/YR, with these flashes allowing for options for the radius of the flash size and well as coloration: https://www.modenc.renegadeprojects.com/Combat_Lighting

Ideally this would be set up as just a Warhead effect type, ie something like CreateLightEffectWarhead or CreateFlashEffectWarhead, with options for adjusting the intensity of the flash, the radius, the color and duration, and maybe positional offset as well.

While this might be able to be recreated in ORA currently using blendmodes, using an engine effect flash seems like a far more elegant and user-friendly solution.

**Comment by alfosua** [comment]
Hey, could I take this one? What should I do to start off the ground? I have read the contributing and code of conduct guidelines. :D

**Comment by alfosua** [comment]
I was playing a few, and so watching gameplays of the original TS. I can't notice where the flash effect takes place. I think there's a flash in almost every weapon, but minimal, though. Could you share me a gif showing what you mean with?

**Comment by alfosua** [comment]
Also, do you mean by adding this effect to every weapon?

**Comment by henskelion** [comment]
The flash would appear when explosions occured, and was sometimes used in place of a dedicated explosion graphic (small debris pieces lacked an explosion SHP and would simply create a light flash wherever they landed, for instance.

I'll try and record a video of TS at some point to show you, though it's a bit difficult since TS is annoying to record with modern software.

Also ideally this flash effect would be something that could be applied to any weapon if so desired. In the original TS it was only used for explosions, not for gunfire (as far as I recall)

**Comment by alfosua** [comment]
Awesome @henskelion. I am exploring through the source code and analyzing how the yaml specs and c# implementations are linked. It'll be great if you could record it, and while it happens, I will be breaking/modding stuff here.

PD: I think that GitHub is already accepting videos in comments now. So you may not need to upload it anywhere than here. Hope it helps.

**Comment by henskelion** [comment]
Getting some footage from TS directly, it appears only certain explosions create the flash effect. I noticed the Orca Transports crashing at the start of GDI mission 2 would causes flashes.
![ts light flash 2](https://user-images.githubusercontent.com/43659306/103181953-c5a3c400-485b-11eb-8b10-e7dd5358ee97.gif)

**Comment by alfosua** [comment]
Ohhhhh, cool effect!

**Comment by alfosua** [comment]
Yeah, it might fit great as well with gunfires. We'd need to make the flash effect sizable to configure its light radius between cases. Configuring its color could be a good plus to increase its reusability. I might name it as `CreateRadialFlashEffectWarhead`.

Let me see what can I do for. Could you assign me this one?

**Comment by pchote** [comment]
To be honest, I don't think this issue is a good fit for a new contributor: it requires a deep understanding of how both the original game and openra's rendering pipelines work.

https://files.zzattack.org/ra2/Ra2DepthBufferSpy.zip can be used to monitor the internal drawing buffers from RA2 (I haven't tested it yet, so i'm not sure if it works for TS), and my understanding is that the original engine had a dedicated buffer for these dynamic light effects that was blended with the rest of the scene in window-space at the end. This can be seen in the animation above as the brightening does not account for the cliffs.

The effect in the animation above is large enough that it could be replicated using our vertex lighting system, but i'm pretty sure that some of the other effects are much smaller (~1 cell) so must be done with pixel-based lighting. I suspect that the best solution here will be to wait until the post-processing shader system is implemented and then handle this similarly to #9778, #12289, and the RGBA-compatible flash/saturation effects: track the drawing regions outside the standard renderable system, and instead apply them using a post-processing shader. This would basically mimic the original game, which is good for recreation accuracy and for performance.
